// @truffle/db/dist/src/resources/bytecodes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bytecodes = void 0;
const logger_1 = require("../logger");
const debug = (0, logger_1.logger)("db:resources:bytecodes");
const graphql_tag_1 = __importDefault(require("graphql-tag"));
const CodeUtils = __importStar(require("@truffle/code-utils"));
exports.bytecodes = {
    names: {
        resource: "bytecode",
        Resource: "Bytecode",
        resources: "bytecodes",
        Resources: "Bytecodes",
        resourcesMutate: "bytecodesAdd",
        ResourcesMutate: "BytecodesAdd"
    },
    createIndexes: [],
    idFields: ["bytes", "linkReferences"],
    typeDefs: (0, graphql_tag_1.default) `
    type Bytecode implements Resource {
      bytes: Bytes!
      linkReferences: [LinkReference!]
      instructions(count: Int): [Instruction!]
    }

    scalar Bytes

    type LinkReference {
      offsets: [ByteOffset!]!
      name: String
      length: Int!
    }

    scalar ByteOffset

    type Instruction {
      opcode: String!
      programCounter: Int!
      pushData: Bytes
    }

    input BytecodeInput {
      bytes: Bytes!
      linkReferences: [LinkReferenceInput!]
    }

    input LinkReferenceInput {
      offsets: [Int!]!
      name: String
      length: Int!
    }
  `,
    resolvers: {
        Bytecode: {
            instructions: {
                resolve({ bytes }, { count }) {
                    return __awaiter(this, void 0, void 0, function* () {
                        const parsed = CodeUtils.parseCode(`0x${bytes}`, {
                            maxInstructionCount: count,
                            attemptStripMetadata: count === undefined
                        });
                        return parsed.map(({ name: opcode, pc: programCounter, pushData }) => ({
                            opcode,
                            programCounter,
                            pushData
                        }));
                    });
                }
            }
        }
    }
};
