// @truffle/dashboard-message-bus-common/dist/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isInvalidateMessage = exports.isDebugMessage = exports.isLogMessage = exports.isDashboardProviderMessage = exports.isCliEventMessage = exports.invalidateMessageType = exports.debugMessageType = exports.logMessageType = exports.dashboardProviderMessageType = exports.cliEventMessageType = void 0;
exports.cliEventMessageType = "cli-event";
exports.dashboardProviderMessageType = "provider";
exports.logMessageType = "log";
exports.debugMessageType = "debug";
exports.invalidateMessageType = "invalidate";
const isCliEventMessage = (message) => {
    return message.type === exports.cliEventMessageType;
};
exports.isCliEventMessage = isCliEventMessage;
const isDashboardProviderMessage = (message) => {
    return message.type === exports.dashboardProviderMessageType;
};
exports.isDashboardProviderMessage = isDashboardProviderMessage;
const isLogMessage = (message) => {
    return message.type === exports.logMessageType;
};
exports.isLogMessage = isLogMessage;
const isDebugMessage = (message) => {
    return message.type === exports.debugMessageType;
};
exports.isDebugMessage = isDebugMessage;
const isInvalidateMessage = (message) => {
    return message.type === exports.invalidateMessageType;
};
exports.isInvalidateMessage = isInvalidateMessage;
