// @trufflesuite/uws-js-unofficial/dist/http-response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpResponse = void 0;
class HttpResponse {
    constructor(res, req) {
        this.response = res;
        this.response.on("close", () => {
            this.abortHandler && this.abortHandler();
        });
        req.on("data", (data) => {
            this.dataHandler && this.dataHandler(data, false);
        });
        req.on("end", () => {
            this.dataHandler && this.dataHandler(Buffer.allocUnsafe(0), true);
        });
    }
    writeContinue() { }
    cork(handler) {
        if (this.response.cork) {
            this.response.cork();
        }
        handler();
        if (this.response.uncork) {
            this.response.uncork();
        }
        return this;
    }
    onWritable(handler) {
        throw new Error("Not implemented");
        return this;
    }
    onAborted(handler) {
        this.abortHandler = handler;
        return this;
    }
    onData(handler) {
        this.dataHandler = handler;
        return this;
    }
    getRemoteAddress() {
        throw new Error("Not implemented");
    }
    getRemoteAddressAsText() {
        throw new Error("Not implemented");
    }
    getProxiedRemoteAddress() {
        throw new Error("Not implemented");
    }
    getProxiedRemoteAddressAsText() {
        throw new Error("Not implemented");
    }
    upgrade(userData, secWebSocketKey, secWebSocketProtocol, secWebSocketExtensions, context) {
        throw new Error("Not implemented");
    }
    write(chunk) {
        return this.response.write(chunk);
    }
    tryEnd(fullBodyOrChunk, totalSize) {
        throw new Error("Not implemented");
    }
    close() {
        if (this.response.socket) {
            this.response.socket.destroy();
        }
        return this;
    }
    getWriteOffset() {
        throw new Error("Not implemented");
    }
    writeStatus(status) {
        this.response.statusCode = parseInt(status.toString());
        return this;
    }
    writeHeader(key, value) {
        this.response.setHeader(key.toString(), value.toString());
        return this;
    }
    pause() {
        throw new Error("Not implemented");
    }
    resume() {
        throw new Error("Not implemented");
    }
    endWithoutBody(reportedContentLength, closeConnection) {
        throw new Error("Not implemented");
    }
    end(data = "", closeConnection = false) {
        if (closeConnection) {
            // Signal that the connection will be closed after completion of the request.
            this.response.setHeader("Connection", "close");
        }
        this.response.end(data);
        if (closeConnection) {
            // also explicitly close the connection to ensure no more data is received from the client on this connection
            this.close();
        }
        return this;
    }
}
exports.HttpResponse = HttpResponse;
