// @trufflesuite/uws-js-unofficial/dist/http-context.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpContext = void 0;
const http_1 = __importDefault(require("http"));
const ws_1 = __importDefault(require("ws"));
const http_router_1 = require("./http-router");
const websocket_1 = require("./websocket");
const UWS_DEFAULT_MAX_PAYLOAD_LENGTH = 16 * 1024;
class HttpContext {
    constructor() {
        this.sockets = new Set();
        this.router = new http_router_1.HttpRouter();
        this.closed = false;
        this.http = null;
        this.wsServer = null;
        this.wsBehaviors = new Map();
    }
    onHttp(method, pattern, handler, upgrade = false) {
        // Todo: This is ugly, fix
        let methods;
        if (method === "*") {
            methods = this.router.methods;
        }
        else {
            methods = [method];
        }
        this.router.add(methods, pattern.toString(), (r) => {
            const user = r.getUserData();
            if (user === null) {
                throw new Error("Could not retrieve user data; this should not happen.");
            }
            user.httpRequest.setYield(false);
            user.httpRequest.setParameters(r.getParameters());
            /* Middleware? Automatically respond to expectations */
            const expect = user.httpRequest.getHeader("expect");
            if (expect != null && expect == "100-continue") {
                user.httpResponse.writeContinue();
            }
            handler(user.httpResponse, user.httpRequest);
            /* If any handler yielded, the router will keep looking for a suitable handler. */
            if (user.httpRequest.getYield()) {
                return false;
            }
            return true;
        }, method === "*"
            ? this.router.LOW_PRIORITY
            : upgrade
                ? this.router.HIGH_PRIORITY
                : this.router.MEDIUM_PRIORITY);
    }
    address() {
        // ganache doesn't support listening on a pipe or Unix domain socket,
        // so `this.http.address()` can never return a string.
        return this.http ? this.http.address() : null;
    }
    onWs(pattern, behavior) {
        // We only need to create the WebSocket Server once, but we shouldn't
        // create it if the user never calls `TemplatedApp.ws(...)`
        if (!this.wsServer) {
            const maxPayloadLength = behavior && "maxPayloadLength" in behavior
                ? behavior.maxPayloadLength
                : UWS_DEFAULT_MAX_PAYLOAD_LENGTH;
            this.wsServer = new ws_1.default.Server({
                // `noServer: true` is necessary for us to be able to call `handleUpgrade`
                // called in `TemplatedApp.upgradeHandler`
                noServer: true,
                maxPayload: maxPayloadLength,
            });
        }
        this.wsBehaviors.set(pattern, behavior);
    }
    handleRequest(req, res) {
        this.router.setUserData(req, res);
        if (!req.method ||
            !req.url ||
            !this.router.route(req.method.toLowerCase(), req.url)) {
            res.destroy();
            return null;
        }
    }
    handleHttpUpgrade(request, socket, head) {
        const pathname = request.url;
        if (this.wsServer !== null) {
            const patterns = this.wsBehaviors.keys();
            for (const pattern of patterns) {
                if (pathname === pattern) {
                    const behavior = this.wsBehaviors.get(pattern);
                    if (typeof behavior.upgrade === "function") {
                        // TODO: custom upgrade handler functionality isn't required
                        // yet in Ganache, so this isn't implemented currently
                        throw new Error("not implemented");
                    }
                    else {
                        this.wsServer.handleUpgrade(request, socket, head, (internalWs) => {
                            const ws = new websocket_1.WebSocket(internalWs);
                            ws.initialize(behavior);
                            this.wsServer.emit("connection", internalWs, request);
                        });
                    }
                    break;
                }
            }
        }
        else {
            socket.end("HTTP/1.1 400 Unexpected server response\n\n");
        }
    }
    listen(host, port, callback) {
        const server = (this.http = http_1.default.createServer(this.handleRequest.bind(this)));
        server.on("connection", (socket) => {
            server.once("close", () => {
                this.sockets.delete(socket);
            });
        });
        server.on("upgrade", this.handleHttpUpgrade.bind(this));
        const handleError = (err) => {
            if (err.syscall === "listen") {
                server.removeListener("error", handleError);
                callback(false);
            }
            else
                throw err;
        };
        server.on("error", handleError);
        server.on("listening", () => server.removeListener("error", handleError));
        const options = { port, exclusive: true };
        if (host) {
            options.host = host.toString();
        }
        server.listen(options, () => {
            this.closed = false;
            callback(this);
        });
    }
    close(cb) {
        this.closed = true;
        for (const socket of this.sockets) {
            socket.destroy();
            this.sockets.delete(socket);
        }
        if (this.wsServer) {
            this.wsServer.close();
        }
        if (this.http) {
            this.http.close(() => {
                this.http = null;
                if (cb) {
                    cb();
                }
            });
        }
        else {
            if (cb) {
                cb();
            }
        }
    }
}
exports.HttpContext = HttpContext;
