// @trufflesuite/uws-js-unofficial/dist/http-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpRequest = void 0;
const url_1 = require("url");
const querystring_1 = require("querystring");
class HttpRequest {
    constructor(request) {
        this.request = request;
        this.ancientHttp = false;
        this.didYield = false;
        this.currentParameters = [0, []];
    }
    isAncient() {
        return this.ancientHttp;
    }
    getYield() {
        return this.didYield;
    }
    /**
     * If you do not want to handle this route
     * @param yield
     */
    setYield(yielded) {
        this.didYield = yielded;
        return this;
    }
    getHeader(lowerCasedHeader) {
        return this.request.headers[lowerCasedHeader] || "";
    }
    getUrl() {
        if (!this.request.url) {
            throw new Error("URL on request is undefined");
        }
        return this.request.url;
    }
    getMethod() {
        if (!this.request.method) {
            throw new Error("Method on request is undefined");
        }
        return this.request.method;
    }
    getCaseSensitiveMethod() {
        throw new Error("Not implemented");
    }
    /**
     * Finds and decodes the URI component.
     * @param key
     */
    getQuery(key) {
        if (!this.request.url) {
            throw new Error("URL on request is undefined");
        }
        const search = new url_1.URL(this.request.url).search;
        if (key != null) {
            return (0, querystring_1.parse)(key)[key] || "";
        }
        else {
            return search;
        }
    }
    setParameters(parameters) {
        this.currentParameters = parameters;
    }
    getParameter(index) {
        if (this.currentParameters[0] < index) {
            return "";
        }
        else {
            return this.currentParameters[1][index];
        }
    }
    forEach(cb) {
        throw new Error("Not implemented");
    }
}
exports.HttpRequest = HttpRequest;
