// @trufflesuite/uws-js-unofficial/dist/templated-app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const http_context_1 = require("./http-context");
class TemplatedApp {
    constructor() {
        this.httpContext = new http_context_1.HttpContext();
    }
    ws(pattern, behavior) {
        this.httpContext.onWs(pattern, behavior);
        this.httpContext.onHttp("get", pattern, (res, req) => {
            const secWebSocketKey = req.getHeader("sec-websocket-key");
            if (secWebSocketKey.length == 24) {
            }
            else {
                req.setYield(true);
            }
        }, true);
        return this;
    }
    get(pattern, handler) {
        this.httpContext.onHttp("get", pattern, handler);
        return this;
    }
    post(pattern, handler) {
        this.httpContext.onHttp("post", pattern, handler);
        return this;
    }
    options(pattern, handler) {
        this.httpContext.onHttp("options", pattern, handler);
        return this;
    }
    del(pattern, handler) {
        this.httpContext.onHttp("delete", pattern, handler);
        return this;
    }
    patch(pattern, handler) {
        this.httpContext.onHttp("patch", pattern, handler);
        return this;
    }
    put(pattern, handler) {
        this.httpContext.onHttp("put", pattern, handler);
        return this;
    }
    head(pattern, handler) {
        this.httpContext.onHttp("head", pattern, handler);
        return this;
    }
    connect(pattern, handler) {
        this.httpContext.onHttp("connect", pattern, handler);
        return this;
    }
    trace(pattern, handler) {
        this.httpContext.onHttp("trace", pattern, handler);
        return this;
    }
    any(pattern, handler) {
        this.httpContext.onHttp("*", pattern, handler);
        return this;
    }
    listen(arg1, arg2, arg3, arg4) {
        let host;
        let port;
        let options;
        let callback;
        if (typeof arg1 === "string") {
            host = arg1;
            port = arg2;
            if (typeof arg3 === "number") {
                options = arg3;
                callback = arg4;
            }
            else {
                options = 0;
                callback = arg3;
            }
        }
        else {
            port = arg1;
            if (typeof arg2 === "number") {
                options = arg2;
                callback = arg3;
            }
            else {
                options = 0;
                callback = arg2;
            }
        }
        this.httpContext.listen(host, port, callback);
        return this;
    }
    publish(topic, message, isBinary, compress) {
        throw new Error("Not implemented");
        return false;
    }
    numSubscribers(topic) {
        throw new Error("Not implemented");
        return 0;
    }
    addServerName(hostname, options) {
        throw new Error("Not implemented");
        return this;
    }
    removeServerName(hostname) {
        throw new Error("Not implemented");
        return this;
    }
    missingServerName(cb) {
        throw new Error("Not implemented");
        return this;
    }
    listen_unix(cb, path) {
        throw new Error("Not implemented");
    }
    domain(domain) {
        throw new Error("Not implemented");
    }
}
exports.default = TemplatedApp;
